pkgs/stdenv/freebsd/unpack-source.sh

```diff
@@ -0,0 +1,41 @@
+unpackCmdHooks+=(_FreeBSDUnpackSource)
+
+_FreeBSDUnpackSource() {
+    local fn="$1"
+    local destination
+
+    if [ -d "$fn" ]; then
+
+        destination="$(stripHash "$fn")"
+
+        if [ -e "$destination" ]; then
+            echo "Cannot copy $fn to $destination: destination already exists!"
+            echo "Did you specify two \"srcs\" with the same \"name\"?"
+            return 1
+        fi
+
+        # We can't preserve hardlinks because they may have been
+        # introduced by store optimization, which might break things
+        # in the build.
+        bsdcp -a -- "$fn" "$destination"
+        chmod -R +w "$destination"
+
+    else
+
+        case "$fn" in
+            *.tar.xz | *.tar.lzma | *.txz)
+                # Don't rely on tar knowing about .xz.
+                xz -d < "$fn" | tar xf - --warning=no-timestamp
+                ;;
```

Comment:
This function is adapted from stdenv/generic/setup.sh _defaultUnpack but made to work with BSD cp, which can handle the nuances of BSD filesystem attributes a lot better. iirc some packages simply won't unpack without overriding the basic unpacker. The verbiage on the comment from the origin seems to have changed since I copied it but it says the same thing.